### ISSUE TITLE
masks2segments function does not always work correctly

### ISSUE THREAD

**Issue by vladoossss** [ultralytics/yolov5#9784]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### Bug

When i use `segment/predict.py --save-txt` to save segmentation results, sometimes i had an error. 

The reason for this was an empty mask (only zeros), which was passed to the input of the masks2segments function. In this regard, `c = cv2.findContours(x, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[0]` produced an empty tuple(due to previous transformations over the mask) and then there was an error.

I made the following changes in 3 places to avoid this:

1. general.py
```
def masks2segments(masks, strategy='largest'):
    # Convert masks(n,160,160) into segments(n,xy)
    segments = []
    for x in masks.int().cpu().numpy().astype('uint8'):
        c = cv2.findContours(x, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[0]
        if c:
            if strategy == 'concat':  # concatenate all segments
                c = np.concatenate([x.reshape(-1, 2) for x in c])
            elif strategy == 'largest':  # select largest segment
                c = np.array(c[np.array([len(x) for x in c]).argmax()]).reshape(-1, 2)
            segments.append(c.astype('float32'))
        else:
            segments.append(0)
    return segments
```

2. segment/predict.py
```
# Segments
if save_txt:
    segments = reversed(masks2segments(masks))
    segments = [scale_segments(im.shape[2:], x, im0.shape).round() if not isinstance(x, int) else 0 for x in segments]
```

3. segment/predict.py
```
# Write results
for j, (*xyxy, conf, cls) in enumerate(reversed(det[:, :6])):
    if save_txt:  # Write to file
        if not isinstance(segments[j], int):
            segj = segments[j].reshape(-1)  # (n,2) to (n*2)
            line = (cls, *segj, conf) if save_conf else (cls, *segj)  # label format
            with open(f'{txt_path}.txt', 'a') as f:
                f.write(('%g ' * len(line)).rstrip() % line + '\n')
        else:
            line = (cls, segments[j], conf) if save_conf else (cls, segments[j])  # label format
            with open(f'{txt_path}.txt', 'a') as f:
                f.write(('%g ' * len(line)).rstrip() % line + '\n')
```

It helps me to avoid an error. 
I understand that my changes are not perfect. The main goal was to point out the problem itself.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@vladoossss thanks! Can you please submit a PR with your proposed changes?

Change #1 doesn't suffice by itself?

**Comment by vladoossss** [comment]
> @vladoossss thanks! Can you please submit a PR with your proposed changes?
> 
> Change #1 doesn't suffice by itself?

Yes, i've submitted new PR, based on this issue.
All 3 changes together worked for me.

**Comment by glenn-jocher** [comment]
@vladoossss good news 😃! Your original issue may now be fixed ✅ in PR #9785. 

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
